src/server/apis/dash-data/handle-incoming-pr-request.ts

```diff
@@ -0,0 +1,371 @@
+import {github} from '../../../utils/github';
+import {userModel, UserRecord} from '../../models/userModel';
+import {getPRLastActivity, LastComment} from '../../utils/get-pr-last-activity';
+import {issueHasNewActivity} from '../../utils/issue-has-new-activity';
+import {DataAPIResponse} from '../api-router/abstract-api-router';
+import * as responseHelper from '../api-router/response-helper';
+import {convertPrFields, convertReviewFields, incomingPrsQuery} from '../dash-data';
+
+type MyReviewFields = reviewFieldsFragment&{commit: {oid: string}};
+
+/**
+ * Handles a response for incoming pull requests.
+ */
+export async function handleIncomingPRRequest(
+    request: express.Request, userRecord: UserRecord):
+    Promise<DataAPIResponse<api.IncomingDashResponse>> {
+  const dashLogin = request.query.login || userRecord.username;
+  const openPrQuery = 'is:open is:pr archived:false';
+  const reviewedQueryString =
+      `reviewed-by:${dashLogin} ${openPrQuery} -author:${dashLogin}`;
+  const reviewRequestsQueryString =
+      `review-requested:${dashLogin} ${openPrQuery}`;
+  const mentionsQueryString = `${reviewedQueryString} mentions:${dashLogin}`;
+
+  const viewerPrsResult = await github().query<IncomingPullRequestsQuery>({
+    query: incomingPrsQuery,
+    variables: {
+      login: dashLogin,
+      reviewRequestsQueryString,
+      reviewedQueryString,
+      mentionsQueryString,
+    },
+    fetchPolicy: 'network-only',
+    context: {
+      token: userRecord.githubToken,
+    }
+  });
+
+  const incomingPrs = [];
+  // In some cases, GitHub will return a PR both in the review request list
+  // and the reviewed list. This ID set is used to ensure we don't show a PR
+  // twice.
+  const prsShown = [];
+
+  // Build a list of mentions.
+  const mentioned: Map<string, api.MentionedEvent> = new Map();
+  for (const item of viewerPrsResult.data.mentions.nodes || []) {
+    if (!item || item.__typename !== 'PullRequest') {
+      continue;
+    }
+    const mention = getLastMentioned(item, dashLogin);
+    if (mention) {
+      mentioned.set(item.id, mention);
+    }
+  }
+
+  let loginRecord: UserRecord|null = null;
+  let lastViewedInfo: {[issue: string]: number}|null = null;
+  if (dashLogin === userRecord.username) {
+    loginRecord = await userModel.getUserRecord(dashLogin);
+    lastViewedInfo = await userModel.getAllLastViewedInfo(dashLogin);
+  }
+
+  // Incoming PRs that I've reviewed.
+  for (const pr of viewerPrsResult.data.reviewed.nodes || []) {
+    if (!pr || pr.__typename !== 'PullRequest') {
+      continue;
+    }
+
+    if (pr.viewerSubscription === 'IGNORED') {
+      continue;
+    }
+
+    // Find relevant review.
+    let relevantReview: MyReviewFields|null = null;
+    if (pr.reviews && pr.reviews.nodes) {
+      // Generated gql-types is missing the proper __type field so a cast is
+      // needed.
+      relevantReview =
+          findMyRelevantReview(pr.reviews.nodes as Array<MyReviewFields|null>);
```

Comment:
This cast may no longer be necessary since I saw that gql-types has been changing.

Reply:
It complains if I remove it so will leave it in for now.